### ISSUE TITLE
feat: LLM-as-a-Judge Analysis Component

### ISSUE THREAD

**Issue by epappas** [epappas/llmtrace#43]
## Executive Summary

Add a third analysis tier to LLMTrace's security ensemble: after regex pattern matching and DeBERTa ML classification, a dedicated LLM performs deep semantic analysis of suspicious content. Operates in two modes -- inline (during request processing) and async (background analysis of stored traces). Supports vLLM (primary), OpenAI API, and Anthropic API backends. Returns structured verdicts that participate in ensemble voting. Persists decisions to storage for the supervised learning pipeline.

## Problem Statement

The current ensemble (regex + DeBERTa) achieves ~84% F1. Key gaps:
1. **Semantic understanding**: Neither regex nor DeBERTa performs genuine semantic reasoning about adversarial intent in context
2. **Novel attacks**: Multi-turn jailbreaks, encoded payloads, context-specific social engineering require reasoning capability
3. **Borderline cases**: Single-detector findings are capped at score 60. An LLM judge can provide a third vote to confirm or suppress
4. **No feedback loop**: No mechanism for continuous improvement based on operational experience

## System Context

```
                          Inline path (optional)
                         +-------------------+
Client -> proxy_handler -> enforcement ->    |
              |             |                v
              |             v           JudgeRouteAction
              |         ActionRouter         |
              |             |                v
              |             v           JudgeClient -> LLM Backend
              |         [decision]          (vLLM/OpenAI/Anthropic)
              |                              |
              v                              v
         background task              JudgeVerdict
              |                              |
              +-> run_security_analysis      v
              +-> ActionRouter::async   JudgeVerdictStore
              |        |               (ClickHouse/PostgreSQL)
              |        v
              |   JudgeRouteAction (enqueue)
              |        |
              v        v
         trace_capture  JudgeWorker (background)
                            |
                            v
                       JudgeClient -> LLM Backend
                            |
                            v
                       JudgeVerdict -> Storage
                            |
                            v
                       ActionRouter::execute_async
                       (secondary actions based on verdict)
```

### Module: `crates/llmtrace-proxy/src/judge.rs`

**Judge verdict:**

```rust
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct JudgeVerdict {
    pub id: Uuid,
    pub trace_id: Uuid,
    pub tenant_id: TenantId,
    pub is_threat: bool,
    pub category: String,           // "prompt_injection", "jailbreak", etc.
    pub confidence: f64,            // 0.0-1.0
    pub security_score: u8,         // 0-100
    pub recommended_action: String, // "allow", "flag", "block"
    pub reasoning: String,
    pub mode: JudgeMode,            // Inline | Async
    pub model_used: String,
    pub latency_ms: u64,
    pub prompt_tokens: Option<u32>,
    pub completion_tokens: Option<u32>,
    pub created_at: DateTime<Utc>,
}
```

**Judge backend trait (multi-backend):**

```rust
#[async_trait]
pub trait JudgeBackend: Send + Sync {
    async fn judge(&self, request: &JudgeRequest) -> Result<JudgeVerdict, JudgeError>;
    fn backend_name(&self) -> &str;
    async fn health_check(&self) -> Result<(), JudgeError>;
}
```

**Backend implementations:**

| Backend | Description | Config |
|---------|-------------|--------|
| `VllmJudgeBackend` | OpenAI-compatible API against self-hosted vLLM | `base_url`, `model`, `max_tokens`, `temperature` |
| `OpenAIJudgeBackend` | OpenAI API (GPT-4o-mini, etc.) | `model`, `max_tokens`, `temperature` |
| `AnthropicJudgeBackend` | Anthropic API (Claude) | `model`, `max_tokens`, `temperature` |

All backends format the `JudgeRequest` into a provider-specific chat completion with a system prompt instructing structured JSON output.

**Judge worker (async background processor):**

```rust
pub struct JudgeWorker {
    receiver: mpsc::Receiver<JudgeRequest>,
    backend: Arc<dyn JudgeBackend>,
    storage: Storage,
    action_router: Option<Arc<ActionRouter>>,
    metrics: Arc<Metrics>,
    config: JudgeConfig,
}

impl JudgeWorker {
    pub async fn run(mut self) {
        while let Some(request) = self.receiver.recv().await {
            // 1. Call judge backend
            // 2. Persist verdict to storage
            // 3. Convert verdict to SecurityFinding
            // 4. Execute secondary actions via ActionRouter
        }
    }
}
```

### Ensemble Integration

Judge verdicts are converted to `SecurityFinding` with `voting_result: "llm_judge"` metadata:

```rust
fn verdict_to_finding(verdict: &JudgeVerdict) -> SecurityFinding {
    // Maps security_score ranges to severity levels
    // Adds judge-specific metadata for ensemble voting
}
```

- **Inline mode**: Judge finding merged into enforcement decision. If ensemble says "single_detector" (capped at 60) and judge confirms, combined score promotes to majority verdict.
- **Async mode**: Judge finding appended to trace in storage. Creates training signal for supervised learning pipeline.

### Storage Schema

**ClickHouse:**
```sql
CREATE TABLE judge_verdicts (
    id UUID, trace_id UUID, tenant_id UUID,
    is_threat Boolean, category String,
    confidence Float64, security_score UInt8,
    recommended_action String, reasoning String,
    mode String, model_used String,
    latency_ms UInt64,
    prompt_tokens Nullable(UInt32),
    completion_tokens Nullable(UInt32),
    created_at DateTime64(3)
) ENGINE = MergeTree()
ORDER BY (tenant_id, created_at)
TTL created_at + INTERVAL 90 DAY;
```

**PostgreSQL:**
```sql
CREATE TABLE judge_verdicts (
    id UUID PRIMARY KEY, trace_id UUID NOT NULL,
    tenant_id UUID NOT NULL, is_threat BOOLEAN NOT NULL,
    category TEXT NOT NULL, confidence DOUBLE PRECISION NOT NULL,
    security_score SMALLINT NOT NULL,
    recommended_action TEXT NOT NULL, reasoning TEXT NOT NULL,
    mode TEXT NOT NULL, model_used TEXT NOT NULL,
    latency_ms BIGINT NOT NULL,
    prompt_tokens INTEGER, completion_tokens INTEGER,
    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
);
```

### FR-01: Multi-Backend Support
- Support vLLM, OpenAI API, and Anthropic API backends
- Backend selection via YAML config
- **AC:** Each backend passes same integration test against mock server

### FR-02: Inline Mode
- When `inline_await: true`, judge called during pre-request enforcement
- Respects timeout; on timeout fail-open (skip verdict)
- **AC:** Known injection triggers inline judge; verdict influences enforcement

### FR-03: Async Mode
- Process requests from mpsc channel in background worker
- Persist verdicts to storage after each judgment
- **AC:** After async analysis, verdict appears in `judge_verdicts` table

### FR-04: Structured Output
- Return `JudgeVerdict` with all required fields
- Malformed LLM JSON output treated as error (no partial verdicts)
- **AC:** Malformed output -> `JudgeError::ParseError` -> logged and skipped

### FR-05: Ensemble Participation
- Verdicts converted to `SecurityFinding` with `voting_result: "llm_judge"` metadata
- Inline: finding included in enforcement decision
- Async: finding appended to trace in storage
- **AC:** Trace in ClickHouse has finding with `voting_result=llm_judge`

### FR-06: Verdict Persistence
- Every verdict stored in `judge_verdicts` (ClickHouse and/or PostgreSQL)
- Includes model name, latency, token counts, reasoning
- **AC:** Query table, all fields populated

### FR-07: Dynamic Enable/Disable
- Disableable at runtime without restart
- When disabled, `JudgeRouteAction` no-ops silently
- **AC:** Disable via config reload; subsequent requests skip judge

### FR-08: Dedicated Model
- Uses separate model from user's upstream request target
- May use same API endpoint with different model name
- **AC:** Config specifies `model: "security-judge-v1"` distinct from upstream

### FR-09: Cost Tracking
- Judge calls tracked separately in cost metrics
- Token usage not conflated with user request tokens
- **AC:** `llmtrace_judge_tokens_total` shows judge-specific consumption

### FR-10: Supervised Learning Pipeline Feed
- Verdicts include context for pipeline: trace_id, category, confidence, reasoning
- **AC:** Join `judge_verdicts` with `traces` returns all training data fields

## Config Schema

```yaml
judge:
  enabled: true
  backend: "vllm"                    # "vllm" | "openai" | "anthropic"
  vllm:
    base_url: "http://localhost:8000"
    model: "security-judge-v1"
    max_tokens: 512
    temperature: 0.1
  openai:
    model: "gpt-4o-mini"             # API key from LLMTRACE_JUDGE_OPENAI_API_KEY
    max_tokens: 512
    temperature: 0.1
  anthropic:
    model: "claude-3-5-haiku-20241022"  # API key from LLMTRACE_JUDGE_ANTHROPIC_API_KEY
    max_tokens: 512
    temperature: 0.1
  worker:
    channel_buffer: 1000
    max_concurrency: 4
    timeout_ms: 30000
  retry:
    max_retries: 2
    backoff_base_ms: 1000
  system_prompt: ""                  # uses built-in default if empty
  min_score_threshold: 30            # only judge traces with score >= this
  persist_verdicts: true
```

## Metrics

| Metric | Type | Labels | Description |
|--------|------|--------|-------------|
| `llmtrace_judge_requests_total` | Counter | `backend`, `mode`, `status` | Total judge invocations |
| `llmtrace_judge_latency_seconds` | Histogram | `backend`, `mode` | Judge call latency |
| `llmtrace_judge_tokens_total` | Counter | `direction`, `backend` | Token consumption |
| `llmtrace_judge_verdicts_total` | Counter | `category`, `recommended_action`, `is_threat` | Verdict distribution |
| `llmtrace_judge_queue_depth` | Gauge | -- | Current async queue depth |
| `llmtrace_judge_verdict_agreement` | Counter | `agreement` | Judge vs ensemble agreement |

## Validation Plan

1. **Unit tests:** `verdict_to_finding()` conversion, JudgeRequest construction, JSON parsing (valid/malformed), min_score_threshold filtering
2. **Integration tests with mock backend:** Each backend against mock HTTP server; retry logic (500 then 200); timeout handling
3. **E2E with real LLM:** Local vLLM with small model; verify verdicts stored in ClickHouse
4. **Ensemble voting:** Regex flags + judge confirms -> majority; regex flags + judge disagrees -> single_detector
5. **Async worker stress:** 1000 requests; verify max_concurrency respected; channel backpressure metrics

## Implementation Sequence

| Phase | Scope | Dependencies |
|-------|-------|-------------|
| 1 | Core types, JudgeBackend trait, VllmJudgeBackend, JSON parsing | None |
| 2 | JudgeWorker, ClickHouse/PostgreSQL migrations, verdict persistence, metrics | Phase 1 |
| 3 | Ensemble integration: verdict_to_finding, inline/async wiring | Phase 1 + Action Router |
| 4 | OpenAIJudgeBackend, AnthropicJudgeBackend, retry/backoff, runtime toggle | Phases 1-3 |

## Known Limitations

- **Latency**: 1-30s per judge call. Inline mode only when latency acceptable; default should be async
- **Cost**: Every call consumes tokens. `min_score_threshold` mitigates by only judging suspicious traces
- **Structured output reliability**: LLMs don't always produce valid JSON. `temperature: 0.1` reduces but doesn't eliminate this
- **Model drift**: Fine-tuned judge may become stale; retraining via pipeline learning required
- **Single judge model**: No judge ensemble; future enhancement
- **No human-in-the-loop**: Judge is autonomous; pipeline learning provides feedback loop

## Risk Assessment

| Risk | Impact | Likelihood | Mitigation |
|------|--------|------------|------------|
| Hallucinated/incorrect verdicts | High | Medium | Ensemble voting (one of three); min_score_threshold; low temperature |
| Judge backend unavailable | Medium | Medium | Fail-open; circuit breaker; fallback to regex+DeBERTa |
| Prompt injection of judge itself | High | Low | Dedicated model with hardened system prompt; analysis text as data |
| Token cost explosion | Medium | Medium | min_score_threshold; max_tokens cap; concurrency limit; separate cost metrics |
| Channel backpressure | Medium | Low | Bounded channel (1000); drop-on-full with metric |

## Dependencies

- **Depends on:** #41 for JudgeRouteAction dispatch; #42 for runtime toggle
- **Required by:** #44 consumes judge verdicts as training labels

**Comment by epappas** [comment]
### Cross-reference from #42 (Dynamic Feature Flags)

As part of #42, a `llm_judge_enabled` field is being added to the `FeatureFlags` struct in `crates/llmtrace-proxy/src/feature_flags.rs` as a **store-only no-op** placeholder — it round-trips through the API and the sidecar `config.runtime.yaml`, but doesn't gate any behavior yet because the LLM Judge backend doesn't exist.

**When picking up this issue, please wire it through:**

1. Add the flag to whatever struct ends up representing the judge config (likely a new `LlmJudgeConfig` block in `ProxyConfig`).
2. Update `FeatureFlags::from_config` / `apply_to_config` in `feature_flags.rs` to map the placeholder field to the real config field instead of the stub.
3. Remove the "llm_judge backend not implemented" warning that `PUT /api/v1/config/features/llm_judge_enabled` currently returns.
4. Add the judge's hot-path gate to read `state.config_handle.snapshot().<...>.llm_judge_enabled` (or equivalent subfield) so it's truly runtime-toggleable per the #42 contract — do not build the judge client once at startup and ignore later flag changes. Per the operating principle from #42: either a feature is genuinely hot-swappable on the request hot path, or it doesn't belong in `FeatureFlags` at all.

**Comment by epappas** [comment]
LLM-as-a-Judge foundation shipped via PR #65, merged to main on 2026-04-21 in commit `287bea4`.

Delivered:
- Three backends: `vllm`, `openai`, `anthropic` + new `deberta` (fast classifier) and `cascade` composite
- Shadow-mode primitive, promotion gate, per-model metrics, fail-open semantics
- Full docs (architecture spec + operator guide + cascade ADR)
- 27-corpus evaluation report (gpt-4o-mini baseline, F1=0.856)
- Live e2e validation with a cascade config

Follow-up research tracked under #86 (cascade), #89 (DeBERTa retrain), #90 (Qwen retrain). E2E framework tracked under #91.
